README.md
[![Pub](https://img.shields.io/pub/v/flutter_aes.svg?style=flat-square)](https://pub.dartlang.org/packages/flutter_aes)
# flutter_aes

一个原生实现的aes加密解密方法，本来原生可以实现，但是数据量大的时候就卡住了

dart的base64和utf8方法不比原生的慢，所以只有加密解密并返回Uint8List
### Example

```dart
import 'package:flutter_aes/flutter_aes.dart';

String _key = "f379e0b661ae4650b19169e4d93665dc";
String _iv = "0000000000000000";
Stopwatch stopwatch = Stopwatch()..start();

var result = base64.encode(await FlutterAes.encrypt(utf8.encode("123"), utf8.encode(_key), utf8.encode(_iv)));
print("encrypt 用时：${stopwatch.elapsed} result:$result");
result = utf8.decode(await FlutterAes.decrypt(
base64.decode(result), utf8.encode(_key), utf8.encode(_iv)));
print("decrypt 用时：${stopwatch.elapsed} result:$result");
```

CHANGELOG.md
## 0.1.0

* 完善描述
## 0.0.1

* 初始化


test/flutter_aes_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_aes/flutter_aes.dart';

void main() {

}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_aes_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


ios/Classes/FlutterAesPlugin.m
#import "FlutterAesPlugin.h"
#import <CommonCrypto/CommonCryptor.h>


@implementation FlutterAesPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"flutter_aes"
                                     binaryMessenger:[registrar messenger]];
    FlutterAesPlugin* instance = [[FlutterAesPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    FlutterStandardTypedData  *key =call.arguments[@"key"];
    FlutterStandardTypedData  *data =call.arguments[@"data"];
    FlutterStandardTypedData  *iv =call.arguments[@"iv"];
    if ([@"decrypt" isEqualToString:call.method]) {
        result([FlutterStandardTypedData typedDataWithBytes:[self operation:data.data keyData: key.data iv:iv.data op:kCCDecrypt]]);
    } else if ([@"encrypt" isEqualToString:call.method]) {
        result([FlutterStandardTypedData typedDataWithBytes:[self operation:data.data keyData: key.data iv:iv.data op:kCCEncrypt]] );
    } else {
        result(FlutterMethodNotImplemented);
    }
}
- (NSData *)operation:(NSData*)contentData keyData:(NSData*)keyData iv:(NSData*)iv op:(CCOperation)op {
    NSUInteger dataLength = [contentData length];
    
    void const *initVectorBytes =  iv.bytes;
    void const *contentBytes = contentData.bytes;
    void const *keyBytes = keyData.bytes;
    
    size_t operationSize = dataLength + kCCBlockSizeAES128;
    void *operationBytes = malloc(operationSize);
    if (operationBytes == NULL) {
        return nil;
    }
    size_t actualOutSize = 0;
    
    CCCryptorStatus cryptStatus = CCCrypt(op,
                                          kCCAlgorithmAES,
                                          kCCOptionPKCS7Padding,
                                          keyBytes,
                                          kCCKeySizeAES256,
                                          initVectorBytes,
                                          contentBytes,
                                          dataLength,
                                          operationBytes,
                                          operationSize,
                                          &actualOutSize);
    
    if (cryptStatus == kCCSuccess) {
        return [NSData dataWithBytesNoCopy:operationBytes length:actualOutSize];
    }
    free(operationBytes);
    operationBytes = NULL;
    return nil;
}
@end


android/src/main/java/com/jlcool/flutter_aes/FlutterAesPlugin.java
package com.jlcool.flutter_aes;

import android.util.Base64;

import java.io.UnsupportedEncodingException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/**
 * FlutterAesPlugin
 */
public class FlutterAesPlugin implements MethodCallHandler {
    /**
     * Plugin registration.
     */
    public static void registerWith(Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_aes");
        channel.setMethodCallHandler(new FlutterAesPlugin());
    }
    @Override
    public void onMethodCall(MethodCall call, Result result) {
        final byte[]  _key = call.argument("key");
        final byte[] _data = call.argument("data");
        final byte[] _iv = call.argument("iv");
        try {
            if (call.method.equals("decrypt")) {
                result.success(cipherOperation(_data, _key, Cipher.DECRYPT_MODE, _iv));
            }   else if (call.method.equals("encrypt")) {
                result.success(cipherOperation(_data, _key, Cipher.ENCRYPT_MODE, _iv));
            } else {
                result.notImplemented();
            }
        } catch (Exception ex) {
            result.success(null);
        }
    }

    public static byte[] cipherOperation(byte[] contentBytes, byte[] keyBytes, int mode,byte[] iv) throws UnsupportedEncodingException, NoSuchAlgorithmException, NoSuchPaddingException, InvalidKeyException, InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException {
        SecretKeySpec secretKey = new SecretKeySpec(keyBytes, "AES");

        IvParameterSpec ivParameterSpec = new IvParameterSpec(iv);
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(mode, secretKey, ivParameterSpec);
        return cipher.doFinal(contentBytes);
    }
}


example/lib/main.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_aes/flutter_aes.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    try {
      String _key = "f379e0b661ae4650b19169e4d93665dc";
      String _iv = "0000000000000000";
      Stopwatch stopwatch = Stopwatch()..start();

      var result = base64.encode(await FlutterAes.encrypt(utf8.encode("123"), utf8.encode(_key), utf8.encode(_iv)));
      print("encrypt 用时：${stopwatch.elapsed} result:$result");
        result = utf8.decode(await FlutterAes.decrypt(
          base64.decode(result), utf8.encode(_key), utf8.encode(_iv)));
      print("decrypt 用时：${stopwatch.elapsed} result:$result");

    } on PlatformException {}

    if (!mounted) return;

    setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
          appBar: AppBar(
            title: const Text('Plugin example app'),
          ),
          body: Center(
              child: FlatButton(
            onPressed: () {
              initPlatformState();
            },
            child: Text("test"),
          ))),
    );
  }
}
